playbooks/adhoc/docker_storage_to_overlay2/docker_storage_to_overlay2.yml

```diff
@@ -0,0 +1,103 @@
+---
+# This playbook changes the Docker storage driver to overlay2.
+#
+# Note, this makes no attempt to migrate existing containers/images.
+# It momentarily stops docker.service and uses "atomic storage reset"
+# to wipe all local Docker storage.
+#
+# To run:
+#
+# ansible-playbook -e 'cli_tag_name=<tag-name>' docker_storage_to_overlay2.yml
+#
+
+- name: Transition Docker storage driver to overlay2
+  hosts: "tag_Name_{{ cli_tag_name }}"
+  user: root
+  connection: ssh
+
+  # For 'ansible_distribution*'
+  gather_facts: yes
+
+  vars:
+    service_names:
+    - "docker.service"
+    optional_service_names:
+    - "atomic-openshift-node.service"
+
+  tasks:
+
+    # XXX Safe to assume atomic is installed?
+
+    - name: Start docker.service for info
+      service:
+        name: docker
+        state: started
+
+    - name: Check if storage driver is already overlay2
+      shell: docker info | grep --ignore-case 'storage driver:.*overlay2'
+      register: storage_driver_check
+      changed_when: False
+      ignore_errors: yes
+
+    - name: Transition storage driver to overlay2
+      when: storage_driver_check.rc == 1
+      block:
+
+        # As of 7.4, SELinux security labels are now supported
+        # on the OverlayFS file system (see RHEL release notes).
+        - name: Check if distribution supports overlayfs (RHEL/CentOS)
+          when: ansible_distribution in ['RedHat', 'CentOS']
+          assert:
+            that: ansible_distribution_version | version_compare('7.4', 'ge')
+            msg: "The OverlayFS file system only works with SELinux in {{ ansible_distribution }} 7.4 or later."
+
+        - name: Check which optional services are available
+          command: systemctl status {{ item }}
+          register: service_available
+          ignore_errors: yes
+          changed_when: False
+          with_items: "{{ optional_service_names }}"
+
+        - name: Add available optional services to service list
+          when: item.rc == 0
+          set_fact:
+            service_names: '{{ service_names + [item.item] }}'
+          with_items: "{{ service_available.results }}"
+
+        - name: Stop services
+          service:
+            name: "{{ item }}"
+            state: stopped
+          with_items: "{{ service_names }}"
+
+        # This step is highly destructive!
+        - name: Remove Docker storage
+          command: atomic storage reset
+
+        - name: Change storage driver to overlay2
+          command: atomic storage modify --driver=overlay2
+
+        - name: Remove any extra storage options for old driver
+          lineinfile:
+            path: /etc/sysconfig/docker-storage-setup
+            regexp: "^{{ item }}="
+            state: absent
+          with_items:
+            - "EXTRA_STORAGE_OPTIONS"
+            - "EXTRA_DOCKER_STORAGE_OPTIONS"  # Deprecated
```

Comment:
We use options like `dm.basesize` that I believe should work on both overlay and devicemapper.

Maybe allow us to pass these options in?

We need this option to be set (and possibly others).

Reply:
`dm.anything` is devicemapper-specific, according to the [docs](https://docs.docker.com/engine/reference/commandline/dockerd/#options-per-storage-driver).

The Docker daemon choked on it when I tried to pass it for `overlay2`:
```
$ systemctl status -l docker
...
... fatal_msg="Error starting daemon: error initializing graphdriver: overlay2: Unknown option dm.basesize\n"
... Failed to start Docker Application Container Engine.
...
```
The only documented option for overlay2 is `overlay2.override_kernel_check`, which we don't need.

Reply:
Ah, interesting. So are we losing the ability to specify basesize?

If so, that's not good. What will keep a single container from eating up all of our docker storage volume?

Reply:
@rhatdan: ^^ Can you answer this?

I'm iterating on your initial devicemapper-to-overlay2 playbook.

Reply:
@twiest Did some Googling on your queston and found [this Moby PR](https://github.com/moby/moby/pull/24771).

The PR enables disk quota limits to be set **per-container** by way of a `--storage-opt size` option when using the overlay2 backend on an XFS filesystem:
e.g. `docker run --storage-opt size=3g -it fedora /bin/bash`

But the catch is the XFS filesystem must be mounted with project quota (`pquota`) support instead of the `noquota` default.  And I don't see a way to inject that mount option by way of `container-storage-setup`.

Does this help at all, or is a limit-per-container a non-starter?

Reply:
@mbarnes that is the functionality we want, but we'd need to ensure that whatever is starting the containers supports that. In this case, that'd mean openshift-node / CRI-O. As far as the pquota mount option, we'll need to have the container storage volume mounted via the fstab I assume, so we can work with that, I think.

@mrunalp  Does CRI-O support setting `--storage-opt size=3g` on container start?   We need a way to limit the amount of space a container can take on the container storage volume.

Reply:
Submitted https://github.com/projectatomic/container-storage-setup/pull/255 to see if we can get the `pquota` option into `container-storage-setup`.

The `docker-storage-setup.service` unit is what actually mounts the volume during boot, and I'd prefer to solve this in the tooling, but we might have to hack `/etc/fstab` as an interim solution.

Reply:
Ah, ok, I didn't realize that.

I agree with you that it'd be better to fix in the tooling.

Reply:
FTR, I spent some time looking at whether specifying a storage limit on the kubelet setting [eviction-hard](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#eviction-thresholds) would be of help here.  But IIUC, that only kicks in when the total available local storage drops below the stated value, and doesn't limit individual containers from growing too big.  Not what we need.

Currently combing through kubelet, dockershim, and CRI-O code to see if there's an insertion point for storage driver options.  The documentation is rather sparse here, but I'm getting the impression this is still an unsolved problem upstream.

Reply:
@twiest : I found what we need!  [Add overlay2.size daemon storage-opt](https://github.com/moby/moby/pull/32977)

But this feature was only just added in June and isn't available where we need it to be.  I emailed you and @rhatdan about it.

Reply:
@mbarnes that's great, good find!